Generate unique id for tensor storage object by observing the week pointer of tensor storage object

shengfukevin opened the pull request:
Summary:
PyTorch execution trace records tensor storage data in the trace. The tensor storage data includes storage id, offset, number of elements, and number of byte for each element. PARAM et-replay uses this information to allocate/free the tensors.
However, the current implementation of generating tensor storage id does not guarantee it is unique. ExecutionTraceObserver maintains a lookup table to map the memory address of the tensor storage object to an unique id. If a new memory address is found, it will be put into that hash table and associate it to a new id.
This implementation does not guarantee the storage object is unique since the memory that the address points to may be released and then re-allocated to a different tensor storage object.


Test Plan: buck2 run mode/opt caffe2/test:test_profiler_cuda -- profiler.test_execution_trace.TestExecutionTraceCUDA

Differential Revision: D75749065

eellison commented on torch/csrc/profiler/standalone/execution_trace_observer.cpp:
```diff
@@ -113,6 +115,65 @@ struct TORCH_API ExecutionTraceObserver { // NOLINT
+    auto iter = storageId.find(raw_data_ptr);
+    if (iter == storageId.end()) {
+      // std::cout << "get_tensor_storage_ID " <<  raw_data_ptr << std::endl;
+      at::DataPtr &data_ptr = t_storage.mutable_data_ptr();
+      auto origin_deleter = data_ptr.get_deleter();
+      bool ret = data_ptr.compare_exchange_deleter(origin_deleter, tensor_storage_deleter);
+      if(ret){
+        ID id = storage_id_++;
```
You can use `getWeakStorageImpl` instead to tell when a tensor has died.

eellison reviewed (commented):
after doing a bit of looking around codebase - you should use getWeakStorageImpl instead.

eellison reviewed (approved):
nice!

ngimel commented:
Please also update PR description

shengfukevin committed:
Generate unique id for tensor storage object by observing the week pointer of tensor storage object (#154859)

Summary:

PyTorch execution trace records tensor storage data in the trace. The tensor storage data includes storage id, offset, number of elements, and number of byte for each element. PARAM et-replay uses this information to allocate/free the tensors.
However, the current implementation of generating tensor storage id does not guarantee it is unique. ExecutionTraceObserver maintains a lookup table to map the memory address of the tensor storage object to an unique id. If a new memory address is found, it will be put into that hash table and associate it to a new id.
This implementation does not guarantee the storage object id is unique since the memory that the address points to may be released and then re-allocated to a different tensor storage object.
This DIFF is to observe the week pointer of tensor storage object in ET. When a tensor storage is deleted, its week pointer will expire. ET saves a map between raw data pointer to the week pointer of the tensor storage object. If the memory address got reused, the week pointer in the map will expire, then the new ID will be generated for it, and the map will be updated with the week pointer to the new tensor storage object.

Test Plan:
buck2 run mode/opt caffe2/test:test_profiler_cuda -- profiler.test_execution_trace.TestExecutionTraceCUDA

Rollback Plan:

Reviewed By: eellison, ngimel

Differential Revision: D75749065

shengfukevin commented:
@pytorchbot merge

shengfukevin commented:
@pytorchmergebot merge main

shengfukevin commented:
@pytorchbot rebase main
